### ISSUE TITLE
Replace Custom GRPO Implementation with VERL Library for Production-Ready Training

### ISSUE THREAD

**Issue by jbarnes850** [Arc-Computer/crm-arena#1]
# Replace Custom GRPO Implementation with VERL Library

## Background

Our current implementation uses a custom-built Group Relative Policy Optimization (GRPO) trainer, which poses significant risks for our startup's proof-of-concept. Custom RL implementations are notoriously error-prone, and accuracy is critical for demonstrating the value of Self-Evolving Curriculum (SEC) learning for business reasoning tasks.

**Risk Assessment:**
- ❌ GRPO advantage calculation bugs could break SEC's core reward mechanism 
- ❌ Reference model synchronization issues
- ❌ Gradient accumulation errors with large batch sizes (256)
- ❌ Numerical instabilities in policy ratios
- ❌ Potential divergence from SEC paper's exact implementation

## Project Context

This is an **open-source proof-of-concept** for our startup that implements Self-Evolving Curriculum (SEC) learning on the CRMArena-Pro benchmark. The implementation must be:

1. **Accurate**: Exactly match the SEC paper's GRPO implementation used by the original authors
2. **Reliable**: Battle-tested framework to avoid training failures
3. **Credible**: Production-ready for demonstrating to investors and partners

From our README.md:

> This repository implements Self-Evolving Curriculum (Chen et al., 2025) for fine-tuning large language models on the CRMArena-Pro benchmark (Huang et al., 2025). The approach formulates curriculum learning as a non-stationary Multi-Armed Bandit problem, using absolute advantages from policy gradient methods as reward signals for curriculum adaptation.

## Solution: Migrate to VERL

The SEC paper authors used **Volcano Engine Reinforcement Learning (VERL) library** as stated in their implementation details:

> "All models are fine-tuned with the GRPO algorithms [44] as implemented in the Volcano Engine Reinforcement Learning (verl) library [45]."

**Benefits:**
- ✅ **Exact paper implementation**: Same library used by SEC authors
- ✅ **Battle-tested**: Production-ready framework by ByteDance team  
- ✅ **Optimized performance**: State-of-the-art throughput and memory efficiency
- ✅ **Active maintenance**: Actively developed with latest features
- ✅ **Community support**: Extensive documentation and examples

### VERL GRPO Configuration
Based on VERL's examples and our current SEC paper configuration:

**Key VERL Parameters for SEC:**
```bash
# Core GRPO settings
algorithm.adv_estimator=grpo
actor_rollout_ref.rollout.n=8                    # SEC paper: 8 rollouts
data.train_batch_size=256                        # SEC paper: critical batch size
actor_rollout_ref.actor.ppo_mini_batch_size=256  # Match global batch size
actor_rollout_ref.actor.use_kl_loss=True         # Enable KL loss
actor_rollout_ref.actor.kl_loss_coef=0.0         # SEC paper: no KL penalty
data.max_prompt_length=1024                      # SEC paper setting
data.max_response_length=4096                    # SEC paper setting
actor_rollout_ref.actor.optim.lr=1e-6            # SEC paper learning rate
algorithm.use_kl_in_reward=False                 # SEC paper: no KL in reward
```

### Current vs Target Architecture

**Current (Custom Implementation):**
```
src/grpo_trainer.py     →  Custom GRPO implementation
src/curriculum.py      →  SEC curriculum logic (keep)
src/data_utils.py       →  Dataset loading (adapt)
train.py               →  Training script (rewrite)
```

**Target (VERL Integration):**
```
src/verl_trainer.py     →  VERL integration wrapper
src/curriculum.py      →  SEC curriculum logic (keep unchanged)
src/verl_data_utils.py  →  VERL-compatible data formatting
train.py               →  VERL training script
```

### Phase 1: Setup and Dependencies
- [ ] **Install VERL**: Add verl to requirements.txt
- [ ] **Verify compatibility**: Test Qwen3-4B model loading with VERL
- [ ] **Environment setup**: Ensure CUDA/PyTorch compatibility

### Phase 2: Data Integration
- [ ] **Data format conversion**: Convert CRMArena-Pro to VERL's expected format (.parquet)
- [ ] **Task categorization mapping**: Integrate our SEC task categorization with VERL data loading
- [ ] **Curriculum sampling**: Implement SEC curriculum arm selection within VERL's data pipeline

### Phase 3: SEC Integration with VERL
- [ ] **VERL wrapper**: Create `VERLSECTrainer` class that integrates SEC curriculum with VERL's GRPO
- [ ] **Advantage extraction**: Hook into VERL's advantage computation to extract SEC rewards (E[|Â|])
- [ ] **Q-value updates**: Implement TD(0) updates for curriculum arms after each training step
- [ ] **Batch sampling**: Implement curriculum-aware batch sampling compatible with VERL

### Phase 4: Configuration Alignment
- [ ] **Match SEC paper config**: Exact parameter alignment with original SEC experiments
- [ ] **Hierarchical MAB**: Integrate 12-arm curriculum structure (4 skills × 3 difficulties)
- [ ] **Hyperparameter mapping**: Convert our YAML config to VERL parameter format

### Phase 5: Validation and Training Readiness
- [ ] **Unit tests**: Comprehensive testing of SEC curriculum integration
- [ ] **Advantage verification**: Verify advantage calculation matches our custom implementation  
- [ ] **Training readiness**: Small validation runs to verify functionality
- [ ] **Documentation update**: Update README and configs for VERL usage

### New Files
- `src/verl_trainer.py` - VERL integration wrapper with SEC curriculum
- `src/verl_data_utils.py` - VERL-compatible data formatting
- `configs/verl_config.yaml` - VERL parameter configuration
- `scripts/convert_data_to_parquet.py` - Data format conversion utility

### Modified Files  
- `train.py` - Replace custom GRPO with VERL training loop
- `requirements.txt` - Add VERL dependencies
- `README.md` - Update installation and usage instructions
- `src/curriculum.py` - Minor adaptations for VERL integration (minimal changes)

### Deprecated Files
- `src/grpo_trainer.py` - Remove custom GRPO implementation
- `configs/config.yaml` - Replace with VERL-compatible configuration

## Success Criteria (Proof of Concept)

1. **VERL integration completes successfully** with Qwen3-4B model loading
2. **SEC curriculum functions correctly** with proper arm selection and Q-value updates
3. **Configuration exactly matches** SEC paper hyperparameters
4. **All validation tests pass** including dataset loading, curriculum sampling, and advantage computation
5. **Training is ready to execute** - all components integrated and validated
6. **Documentation is complete** for VERL setup and usage

## Expected Benefits

- **Reduced risk**: Production-ready framework eliminates custom implementation bugs
- **Improved credibility**: Using the same framework as the original SEC paper
- **Better performance**: VERL's optimizations for throughput and memory efficiency
- **Future-proofing**: Access to latest RL algorithm implementations and optimizations
- **Community support**: Extensive documentation and active development

## Timeline Estimate (Proof of Concept)

- **Phase 1-2**: 1-2 days (setup and data integration)
- **Phase 3-4**: 2-3 days (SEC integration and configuration)  
- **Phase 5**: 1 day (validation and training readiness)

**Total: 4-6 days for complete migration and training readiness**

## Technical Dependencies

```
# Core requirements
verl>=0.3.0
torch>=2.0.0
transformers>=4.36.0
datasets>=2.14.0
vllm>=0.8.2  # For rollout generation
wandb>=0.15.0  # For experiment tracking
```

## References for Implementation

- [VERL Repository](https://github.com/volcengine/verl)
- [VERL GRPO Examples](https://github.com/volcengine/verl/tree/main/examples/grpo_trainer)
- [Qwen3-4B VERL Script](https://github.com/volcengine/verl/blob/main/examples/grpo_trainer/run_qwen3-8b.sh)
- [SEC Paper Implementation Details](docs/self-evolving-curriculum.txt)
- [VERL Documentation](https://verl.readthedocs.io/en/latest/)

## Current Implementation Context

**Key files to understand:**
- `src/grpo_trainer.py` - Current custom GRPO implementation to be replaced
- `src/curriculum.py` - SEC curriculum logic (keep unchanged)
- `src/data_utils.py` - Dataset loading and sampling logic
- `configs/config.yaml` - Current configuration (batch_size=256, lr=1e-6, rollouts=8)
- `README.md` - Project overview and current setup instructions

**Current configuration aligned with SEC paper:**
```yaml
training:
  batch_size: 256  # SEC paper critical batch size
  learning_rate: 0.000001  # SEC paper learning rate (1e-6)
  rollouts: 8  # SEC paper rollouts for advantage estimation
  kl_penalty_weight: 0.0  # No KL penalty in SEC paper
```

---

**Assignee**: Claude (AI Assistant)
**Priority**: High  
**Labels**: enhancement, critical, production-ready

This migration is essential for the credibility and reliability of our SEC proof-of-concept. Using the same framework as the original paper authors eliminates implementation risk and ensures our results are directly comparable to published benchmarks.
